File: frontend/src/api/tutor.ts
import { api } from "./client";

export interface TutorAskRequest {
  subject: string;
  topic: string;
  question: string;
}

export interface TutorAskResponse {
  answer: string;
  examples: string[];
  related_topics: string[];
  difficulty_assessment: string;
  suggested_resources: string[];
}

export interface TutorExplainRequest {
  subject: string;
  topic: string;
  mode: "simple" | "advanced";
}

export interface TutorExplainResponse {
  explanation: string;
  examples: string[];
  analogies: string[];
  formulas: string[];
  code_examples: string[];
  key_takeaways: string[];
}

export interface QuizQuestion {
  id: number;
  question: string;
  options: Record<string, string>;
  correct_answer: string;
  explanation: string;
  difficulty: string;
}

export interface TutorQuizRequest {
  subject: string;
  topic: string;
  question_count: number;
  difficulty: string;
}

export interface TutorQuizResponse {
  questions: QuizQuestion[];
  total_questions: number;
}

export interface QuizAnswerPayload {
  question_id: number;
  question: string;
  selected_answer: string;
  correct_answer: string;
}

export interface TutorEvaluateQuizRequest {
  subject: string;
  topic: string;
  answers: QuizAnswerPayload[];
}

export interface PerQuestionFeedback {
  question_id: number;
  correct: boolean;
  feedback: string;
}

export interface TutorEvaluateQuizResponse {
  score: number;
  total: number;
  percentage: number;
  per_question_feedback: PerQuestionFeedback[];
  weak_topics: string[];
  strong_topics: string[];
  recommendations: string[];
}

export interface StudyPlanDay {
  day: number;
  topics: string[];
  duration_hours: number;
  activities: string[];
  resources: string[];
}

export interface TutorStudyPlanRequest {
  subject: string;
  exam_date: string;
  duration_days: 7 | 30;
}

export interface TutorStudyPlanResponse {
  plan: StudyPlanDay[];
  total_hours: number;
  exam_strategy: string;
  prerequisites: string[];
  tips: string[];
}

export interface TutorHistoryItem {
  id: number;
  session_type: string;
  subject: string | null;
  topic: string | null;
  question: string | null;
  answer: Record<string, any> | null;
  created_at: string;
}

export interface TutorHistoryResponse {
  history: TutorHistoryItem[];
}

export async function askDoubt(data: TutorAskRequest): Promise<TutorAskResponse> {
  const res = await api.post("/ai/tutor/ask", data);
  return res.data;
}

export async function explainTopic(data: TutorExplainRequest): Promise<TutorExplainResponse> {
  const res = await api.post("/ai/tutor/explain", data);
  return res.data;
}

export async function generateQuiz(data: TutorQuizRequest): Promise<TutorQuizResponse> {
  const res = await api.post("/ai/tutor/quiz", data);
  return res.data;
}

export async function evaluateQuiz(data: TutorEvaluateQuizRequest): Promise<TutorEvaluateQuizResponse> {
  const res = await api.post("/ai/tutor/evaluate-quiz", data);
  return res.data;
}

export async function createStudyPlan(data: TutorStudyPlanRequest): Promise<TutorStudyPlanResponse> {
  const res = await api.post("/ai/tutor/study-plan", data);
  return res.data;
}

export async function getTutorHistory(limit: number = 50): Promise<TutorHistoryResponse> {
  const res = await api.get(`/ai/tutor/history?limit=${limit}`);
  return res.data;
}

File: backend/app/schemas/tutor.py
from datetime import datetime
from typing import List, Optional, Literal

from pydantic import BaseModel, Field


class TutorAskRequest(BaseModel):
    subject: str
    topic: str
    question: str


class TutorAskResponse(BaseModel):
    answer: str
    examples: List[str]
    related_topics: List[str]
    difficulty_assessment: str
    suggested_resources: List[str]


class TutorExplainRequest(BaseModel):
    subject: str
    topic: str
    mode: Literal["simple", "advanced"] = "simple"


class TutorExplainResponse(BaseModel):
    explanation: str
    examples: List[str]
    analogies: List[str]
    formulas: List[str]
    code_examples: List[str]
    key_takeaways: List[str]


class QuizQuestion(BaseModel):
    id: int
    question: str
    options: dict[str, str]
    correct_answer: str
    explanation: str
    difficulty: str


class TutorQuizRequest(BaseModel):
    subject: str
    topic: str
    question_count: int = Field(default=5, ge=1, le=20)
    difficulty: str = "medium"


class TutorQuizResponse(BaseModel):
    questions: List[QuizQuestion]
    total_questions: int


class QuizAnswer(BaseModel):
    question_id: int
    question: str
    selected_answer: str
    correct_answer: str


class TutorEvaluateQuizRequest(BaseModel):
    subject: str
    topic: str
    answers: List[QuizAnswer]


class PerQuestionFeedback(BaseModel):
    question_id: int
    correct: bool
    feedback: str


class TutorEvaluateQuizResponse(BaseModel):
    score: int
    total: int
    percentage: float
    per_question_feedback: List[PerQuestionFeedback]
    weak_topics: List[str]
    strong_topics: List[str]
    recommendations: List[str]


class StudyPlanDay(BaseModel):
    day: int
    topics: List[str]
    duration_hours: float
    activities: List[str]
    resources: List[str]


class TutorStudyPlanRequest(BaseModel):
    subject: str
    exam_date: str
    duration_days: Literal[7, 30] = 7


class TutorStudyPlanResponse(BaseModel):
    plan: List[StudyPlanDay]
    total_hours: float
    exam_strategy: str
    prerequisites: List[str]
    tips: List[str]


class TutorHistoryItem(BaseModel):
    id: int
    session_type: str
    subject: Optional[str]
    topic: Optional[str]
    question: Optional[str]
    answer: Optional[dict]
    created_at: datetime


class TutorHistoryResponse(BaseModel):
    history: List[TutorHistoryItem]
